### PR TITLE
Document builtin topics and blocklist [13993]

### PR DESCRIPTION
**Pull request by rsanchez15**
Merge to [feature/dynamic-topic-discovery](https://github.com/eProsima/DDS-Router/tree/feature/dynamic-topic-discovery).

### CODE REVIEW

File: docs/rst/user_manual/configuration.rst [hunk 1]
```diff
@@ -14,16 +14,22 @@ and :term:`Participants <Participant>` configurations.
...
-
-    The |ddsrouter| topic discovery module is a work in progress.
-    Thus, the functionality regarding Topic filtering is still in its early stages.
-
-The |ddsrouter| requires a list of allowed :term:`Topics<Topic>`.
-The data transmitted under these topics will be the data relayed by the |ddsrouter|.
-The YAML configuration file must contain an ``allowlist`` tag which is a vector (``[]``).
-This vector elements contains the Topics that will be forwarded (subsequently the |ddsrouter| creates a
-:term:`Writers<DataWriter>` and :term:`Readers<DataReader>` for each topic for each Participant).
+|ddsrouter| allows filtering of DDS :term:`Topics<Topic>`, that is, it allows defining which DDS Topics are going to be
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:17]:
```suggestion
|ddsrouter| allows filtering of DDS :term:`Topics<Topic>`, that is, it allows to define which DDS Topics are going to be
```

File: docs/rst/user_manual/configuration.rst [hunk 2]
```diff
@@ -14,16 +14,22 @@ and :term:`Participants <Participant>` configurations.
...
+relayed by the application.
+This way, it is possible to define a set of rules in |ddsrouter| to filter those data that the user do not want to be
+forwarded. It is not mandatory to define a rule set in the configuration file.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:20]:
```suggestion
This way, it is possible to define a set of rules in |ddsrouter| to filter those data samples the user does not wish to forward. It is not mandatory to define such set of rules in the configuration file.
```

File: docs/rst/user_manual/configuration.rst [hunk 3]
```diff
@@ -14,16 +14,22 @@ and :term:`Participants <Participant>` configurations.
...
+In this case, a DDS Router will forward all the data published under the topics that it automatically discovers within
+the DDS network to which it connects.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:22]:
```suggestion
In this case, a DDS Router will forward all the data published under the topics that it automatically discovers within
the DDS network to which it is connected.
```

File: docs/rst/user_manual/configuration.rst [hunk 4]
```diff
@@ -14,16 +14,22 @@ and :term:`Participants <Participant>` configurations.
...
+
+To define these data filtering rules based on the Topics to which they belong, three lists are available:
+
+* Builtin topics list (``builtin-topics``)
+* Allowed topics list (``allowlist``)
+* Block topics list (``blocklist``)
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:28]: Maybe better explain `builtin-topics` the last one, and let's try to use this ordering everywhere else.

File: docs/rst/user_manual/configuration.rst [hunk 5]
```diff
@@ -48,26 +54,86 @@ Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with o
...
-.. todo:
+.. note::
+
+    Tags ``allowlist``, ``blocklist`` and ``builtin-topics`` must be at yaml base level (it must not be inside any
+    other tag).
+
+Allow topic list (``allowlist``)
+--------------------------------
+This is the list of topics that |ddsrouter| will forward, i.e. the data published under the topics defined in
+the ``allowlist`` will be relayed by |ddsrouter|.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:65]:
```suggestion
This is the list of topics that |ddsrouter| will forward, i.e. the data published under the topics matching the expressions in
the ``allowlist`` will be relayed by |ddsrouter|.
```

File: docs/rst/user_manual/configuration.rst [hunk 6]
```diff
@@ -48,26 +54,86 @@ Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with o
...
+
+
+Block topic list (``blocklist``)
+--------------------------------
+This is the list of topics that the |ddsrouter| will block, that is, all data published under the topics defined in
+the ``blocklist`` will be discarded by the |ddsrouter| and therefore will not be relayed.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:71]:
```suggestion
This is the list of topics that the |ddsrouter| will block, that is, all data published under the topics matching the expressions in
the ``blocklist`` will be discarded by the |ddsrouter| and therefore will not be relayed.
```

File: docs/rst/user_manual/configuration.rst [hunk 7]
```diff
@@ -48,26 +54,86 @@ Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with o
...
+
+This list takes precedence over the ``allowlist``.
+If there is a topic defined in the ``allowlist`` and in the ``blocklist``, the ``blocklist`` takes precedence, causing
+the data of this topic to be discarded.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:75]:
```suggestion
If a topic matches an expression both in the ``allowlist`` and in the ``blocklist``, the ``blocklist`` takes precedence, causing
the data under this topic to be discarded.
```

File: docs/rst/user_manual/configuration.rst [hunk 8]
```diff
@@ -48,26 +54,86 @@ Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with o
...
+
+
+Builtin topics list (``builtin-topics``)
+----------------------------------------
+
+|ddsrouter| has a mechanism to automatically detect which topics are being used in a DDS network.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:81]:
```suggestion
|ddsrouter| includes a mechanism to automatically detect which topics are being used in a DDS network.
```

File: docs/rst/user_manual/configuration.rst [hunk 9]
```diff
@@ -48,26 +54,86 @@ Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with o
...
+By automatically detecting these topics, a |ddsrouter| creates internal DDS :term:`Writers<DataWriter>`
+and :term:`Readers<DataReader>` for each topic and for each Participant in order to relay the data published on each
+discovered topic.
+
+The discovery phase of the network topics can be accelerated by using the builtin topic list (``builtin-topics``).
+By defining topics in this list, the DDS router will create the DataWriters and DataReaders for these topics without
+the need to have previously discovered them.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:88]:
```suggestion
By defining topics in this list, the DDS router will create the DataWriters and DataReaders for these topics without waiting for them to be discovered.
```

File: docs/rst/user_manual/configuration.rst [hunk 10]
```diff
@@ -48,26 +54,86 @@ Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with o
...
+This example shows how the |ddsrouter| is initially configured to forward the ``rt/chatter`` topic (default ROS 2
+topic for ``talker`` and ``listener``) with type name ``std_msgs::msg::dds_::String_``, while the rest of the
+topics in the DDS network are expected to be dynamically discovered.
+Additionally, a rule is defined to block only the ``HelloWorldTopic`` topic.
 
 .. code-block:: yaml
 
-    allowlist:
+    blocklist:
+      - name: "HelloWorldTopic"
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:108]: Remove quotes everywhere.

Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:108]:
```suggestion
      - name: *
```

Comment (juanlofer-eprosima) [reply docs/rst/user_manual/configuration.rst:108]: Just a suggestion in order to include wildcard expressions in the example, maybe it is more interesting for an user to block all topics of a certain type.

File: docs/rst/user_manual/configuration.rst [hunk 11]
```diff
@@ -48,26 +54,86 @@ Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with o
...
+
+    builtin-topics:
       - name: "rt/chatter"
         type: "std_msgs::msg::dds_::String_"
+
+Allowlist and blocklist collision
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the following example, the ``HelloWorldTopic`` topic is in both the ``allowlist`` and the
+``blocklist`` so according to the ``blocklist`` preference rule, this topic is blocked.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/configuration.rst:119]:
```suggestion
``blocklist``,  so according to the ``blocklist`` preference rule this topic is blocked.
```

File: docs/rst/user_manual/user_interface.rst [hunk 12]
```diff
@@ -84,8 +84,9 @@ It shows the usage information of the application.
     General options:
     -h --help         Print this help message.
     -c --config-path  Path to the Configuration File (yaml format) [Default: ./DDS_ROUTER_CONFIGURATION.yaml].
-    -r --reload-time  Time period in seconds to reload configuration file. This is needed when FileWatcher functionality is not available (e.g. config file is a symbolic link).
-                        Value 0 does not reload file. [Default: 0].
+    -r --reload-time  Time period in seconds to reload configuration file. This is needed when File Watcher
+                      functionality is not available (e.g. config file is a symbolic link).
+                      Value 0 does not reload file. [Default: 0].
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/user_interface.rst:89]: Changed on purpose?

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: README.md [hunk 13]
```diff
@@ -35,6 +35,8 @@ Following are some of the key features of *eProsima DDS Router*:
   networks.
 * **Topic allowlisting**: it is possible to configure a *DDS Router* to forward just the user data belonging to a
   topic specified by the user.
+* **Dynamic topic discovery**: underlying entities required for communication under a specific topic are created only
```
Comment (jparisu) [inline README.md:38]: I would no say so.
The important part of dynamic module is that user does not need to know topics and types beforehand, not to minimize resources (which is not true)

File: docs/resources/getting_started/client-ddsrouter.yaml [hunk 14]
```diff
@@ -3,7 +3,6 @@
 allowlist:
   - name: Square
     type: ShapeType
```
Comment (jparisu) [inline docs/resources/getting_started/client-ddsrouter.yaml:5]: I would remove type as well. The topic is known by the user (it decides the figure shape), but it does not know the type.

File: docs/resources/getting_started/server-ddsrouter.yaml [hunk 15]
```diff
@@ -3,7 +3,6 @@
 allowlist:
   - name: Square
     type: ShapeType
```
Comment (jparisu) [inline docs/resources/getting_started/server-ddsrouter.yaml:5]: Same, I would remove it

File: docs/rst/formalia/titlepage.rst [hunk 16]
```diff
@@ -39,6 +39,8 @@ Following are some of the key features of *eProsima DDS Router*:
      networks.
   5. **Topic allowlisting**: it is possible to configure a *DDS Router* to forward just the user data belonging to a
      topic specified by the user.
+  6. **Dynamic topic discovery**: underlying entities required for communication under a specific topic are created only
```
Comment (jparisu) [inline docs/rst/formalia/titlepage.rst:42]: same as in README

File: docs/rst/getting_started/usage_example.rst [hunk 17]
```diff
@@ -48,18 +48,15 @@ default settings).
 Router configuration
 ====================
 A configuration file is all that is required in order to run a |ddsrouter| instance. In a nutshell, each router will
-only forward messages if their associated topics are contained in its ``allowlist``. A ``blocklist`` can also be
-specified, on its own or in addition to an ``allowlist``, but we will not be covering this here.
+only forward messages if their associated topics match the filters contained in its ``allowlist``. A ``blocklist``
```
Comment (jparisu) [inline docs/rst/getting_started/usage_example.rst:51]:
```suggestion
forward messages if their associated topics match the filters contained in its ``allowlist``. A ``blocklist``
```

File: docs/rst/user_manual/configuration.rst [hunk 18]
```diff
@@ -43,19 +50,66 @@ Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with o
...
+causing the data under this topic to be discarded.
+
+
+Builtin topics list (``builtin-topics``)
+----------------------------------------
+
+|ddsrouter| includes a mechanism to automatically detect which topics are being used in a DDS network.
+By automatically detecting these topics, a |ddsrouter| creates internal DDS :term:`Writers<DataWriter>`
+and :term:`Readers<DataReader>` for each topic and for each Participant in order to relay the data published on each
+discovered topic.
```
Comment (jparisu) [inline docs/rst/user_manual/configuration.rst:90]: I would move this to the main Topic filtering paragraph, I think it is main information.
If not this, at least mention it above

File: docs/rst/user_manual/configuration.rst [hunk 19]
```diff
@@ -65,9 +119,36 @@ with type name ``HelloWorld``.
       - name: HelloWorldTopic
         type: HelloWorld
 
-.. note::
+    blocklist:
+      - name: "*"
+        type: HelloWorld
```
Comment (jparisu) [inline docs/rst/user_manual/configuration.rst:124]: Use real cases to exemplify. rq and rr are the name for topics in services in ROS2 
```suggestion
      - name: "rq/*"
      - name: "rr/*"
```

File: docs/rst/user_manual/user_interface.rst [hunk 20]
```diff
@@ -6,7 +6,7 @@
 User Interface
 ##############
 
-|eddsrouter| is an application executed from command line.
+|eddsrouter| is an user application executed from command line and configured through a YAML configuration file.
```
Comment (jparisu) [inline docs/rst/user_manual/user_interface.rst:9]:
```suggestion
|eddsrouter| is a user application executed from command line and configured through a YAML configuration file.
```

File: docs/rst/user_manual/user_interface.rst [hunk 21]
```diff
@@ -84,8 +84,9 @@ It shows the usage information of the application.
     General options:
     -h --help         Print this help message.
     -c --config-path  Path to the Configuration File (yaml format) [Default: ./DDS_ROUTER_CONFIGURATION.yaml].
-    -r --reload-time  Time period in seconds to reload configuration file. This is needed when FileWatcher functionality is not available (e.g. config file is a symbolic link).
-                        Value 0 does not reload file. [Default: 0].
+    -r --reload-time  Time period in seconds to reload configuration file. This is needed when File Watcher
```
Comment (jparisu) [inline docs/rst/user_manual/user_interface.rst:87]: I'd leave this text exactly as the exit format in tool --help

File: docs/rst/user_manual/configuration.rst [hunk 22]
```diff
@@ -43,19 +50,66 @@ Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with o
...
+.. note::
+
+    Placing quotation marks around values in a YAML file is generally optional. However, values containing wildcard
+    characters must be enclosed by single or double quotation marks.
+
+Allow topic list (``allowlist``)
+--------------------------------
+This is the list of topics that |ddsrouter| will forward, i.e. the data published under the topics matching the
+expressions in the ``allowlist`` will be relayed by |ddsrouter|.
+
```
Comment (jparisu) [inline docs/rst/user_manual/configuration.rst:72]: Add a note (or warning) saying that an empty allowlist allow every topic

**Review by jparisu — CHANGES_REQUESTED** [review]

File: docs/rst/user_manual/configuration.rst (hunk 18 above)
Comment (juanlofer-eprosima) [reply docs/rst/user_manual/configuration.rst:90]: It is actually in the first paragraph.

File: docs/rst/user_manual/configuration.rst (hunk 19 above)
Comment (jparisu) [reply docs/rst/user_manual/configuration.rst]: The idea was to filter them, so the whole .yaml refers to a real use case.

File: docs/rst/user_manual/configuration.rst [hunk 23]
```diff
@@ -10,20 +10,27 @@ A |ddsrouter| is configured by a *.yaml* configuration file.
...
 Topic Filtering
 ===============
 
-.. note::
+|ddsrouter| allows filtering of DDS :term:`Topics<Topic>`, that is, it allows to define which DDS Topics are going to be
+relayed by the application.
+This way, it is possible to define a set of rules in |ddsrouter| to filter those data samples the user does not wish to
+forward. It is not mandatory to define such set of rules in the configuration file.
+In this case, a DDS Router will forward all the data published under the topics that it automatically discovers within
+the DDS network to which it connects.
```
Comment (jparisu) [inline docs/rst/user_manual/configuration.rst:23]: I will try to refactor these sentences.
First, separate the explanation about the configuration and the use of topic filtering.
Second, saying that data samples are filtered is incorrect, it should say DDS entities are created/activated in those topics.
Third, make it clearer that when a new topic is discovered in the network, Router will create entities to forward it if allowed.

**Review by jparisu — CHANGES_REQUESTED** [review]
Small changes remain

File: docs/rst/user_manual/configuration.rst (hunk 23 above)
Comment (rsanchez15) [reply docs/rst/user_manual/configuration.rst:23]: The use of these phrases without reference to the creation of internal entities is intended to be more understandable to the user. Therefore I would leave them as they are.

**Review by jparisu — APPROVED** [review]
LGTM

**Pull request merged**